tlslite/keyexchange.py

```diff
@@ -734,3 +718,88 @@ def makeClientKeyExchange(self):
         cke = super(SRPKeyExchange, self).makeClientKeyExchange()
         cke.createSRP(self.A)
         return cke
+
+
+class RawDHKeyExchange(object):
+    """
+    Abstract class for performing Diffe-Hellman key exchange.
+
+    Provides a shared API for X25519, ECDHE and FFDHE key exchange.
+    """
+
+    def __init__(self, group, version):
+        """
+        Set the parameters of the key exchange
+
+        Sets group on which the KEX will take part and protocol version used.
+        """
+        self.group = group
+        self.version = version
+
+    def get_random_private_key(self):
+        """
+        Generate a random value suitable for use as the private value of KEX.
+        """
+        raise NotImplementedError("Abstract class")
+
+    def calc_public_value(self, private):
+        """Calculate the public value from the provided private value."""
+        raise NotImplementedError("Abstract class")
+
+    def calc_shared_key(self, private, peer_share):
+        """Calcualte the shared key given our private and remote share value"""
+        raise NotImplementedError("Abstract class")
+
+
+class FFDHKeyExchange(RawDHKeyExchange):
+    """Implemenation of the Finite Field Diffie-Hellman key exchange."""
+
+    def __init__(self, group, version, generator=None, prime=None):
+        super(FFDHKeyExchange, self).__init__(group, version)
+        if prime and group:
+            raise ValueError("Can't set the RFC7919 group and custom params"
+                             " at the same time")
+        if group:
```

Comment:
Can't this group ID to DH parameter mapping be moved to the caller or a separate function?  In both cases where ```group``` is set or ```None```, self.prime and self.generator is set once in the constructor, so I would rather make them as a mandatory argument of the constructor:
```python
class FFDHKeyExchange(RawDHKeyExchange):
    def __init__(self, generator, prime):
        ...
```

Similarly, ```ECDHKeyExchange``` can also be:
```python
class ECDHKeyExchange(RawDHKeyExchange):
    def __init__(self, curve, ...):
        ...
class 
```

When RFC7919 mapping is needed, you could define a factory method, something like
```python
def createKeyExchange(group_name, version, ...):
    if GroupName.ffdhe2048 <= group_name <= GroupName.ffdhe8192:
        generator, prime = RFC7919_GROUPS[group_name-GroupName.ffdhe2048]
        return FFDHKeyExchange(generator, prime)
    elif ... ECDH case ...:
```

Reply:
That's essentially how I did this later:
https://github.com/tomato42/tlslite-ng/pull/174/files#diff-51b5feff926add26aec88a6e6e00a78dR927

I was mostly thinking to make the TLS 1.3 code cleaner, as that will be the primary use case, and make the old code "suffer" by requiring it to pass the `None` as the group id